bootstrap/bootstrap.go

```diff
@@ -1792,31 +1796,54 @@ func (b *Bootstrap) ignoredEnv() []string {
 	return ignored
 }
 
-// Check the redaction config and create a redactor if necessary - may return
-// nil if there's nothing to redact.
-// The redactor is returned so the caller can `defer redactor.Flush()`
-func (b *Bootstrap) setupRedactor() *Redactor {
+// setupRedactor wraps shell output and logging in Redactor if any redaction
+// is necessary based on RedactedVars configuration and the existence of
+// matching environment vars.
+// RedactorMux (possibly empty) is returned so the caller can `defer redactor.Flush()`
+func (b *Bootstrap) setupRedactor() RedactorMux {
 	if experiments.IsEnabled("output-redactor") {
 		b.shell.Commentf("Using output-redactor experiment 🧪")
 	} else {
 		return nil
 	}
-
 	valuesToRedact := getValuesToRedact(b.shell, b.Config.RedactedVars, b.shell.Env.ToMap())
+	var mux RedactorMux
 
-	// If the shell Writer is already a Redactor, don't layer another Redactor
-	// on top of it
+	// If the shell Writer is already a Redactor, reset the values to redact.
 	if redactor, ok := b.shell.Writer.(*Redactor); ok {
-		// Still need to reset the values when re-using the Redactor
 		redactor.Reset(valuesToRedact)
-		return redactor
+		mux = append(mux, redactor)
+	} else if len(valuesToRedact) == 0 {
+		// skip
+	} else {
+		redactor := NewRedactor(b.shell.Writer, "[REDACTED]", valuesToRedact)
+		b.shell.Writer = redactor
+		mux = append(mux, redactor)
 	}
-	if len(valuesToRedact) == 0 {
-		return nil
+
+	// If the shell.Logger is already a redacted WriterLogger, reset the values to redact.
+	// (maybe there's a better way to do two levels of type assertion? ...
+	// shell.Logger may be a WriterLogger, and its Writer may be a Redactor)
+	var shellWriterLogger *shell.WriterLogger
+	var shellLoggerRedactor *Redactor
+	if logger, ok := b.shell.Logger.(*shell.WriterLogger); ok {
+		shellWriterLogger = logger
+		if redactor, ok := logger.Writer.(*Redactor); ok {
+			shellLoggerRedactor = redactor
+		}
```

Comment:
👆🏼 I didn't like writing this code, and I don't like reading it. It's the closest and simplest I could come to the equivalent of the previous code, but with two different redactors, and with one of them only applicable to a certain concrete type of the shell logger.

It determines whether `b.shell.Logger` is a `shell.WriterLogger` (which it is by default outside of tests), because that's the implementation of shell's `type Logger interface { … }` that has an `io.Writer` which can be wrapped in a redactor.

And if it _is_ a `shell.WriterLogger`, it checks whether it's already wrapped in a redactor, in which case the subsequent code `Reset()`s it instead of wrapping it (just like the previous code did with the original output redactor).